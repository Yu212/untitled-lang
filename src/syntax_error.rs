use rowan::TextRange;

#[derive(Debug)]
pub struct SyntaxError {
    pub message: String,
    pub range: TextRange,
}
