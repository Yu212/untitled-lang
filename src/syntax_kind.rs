#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Error,
    Whitespace,
    Eof,
    EqEq,
    Neq,
    Comma,
    Semicolon,
    Colon,
    Ident,
    TypedIdent,
    TypeSpec,
    IdentTypeSpec,
    ArrayTypeSpec,
    LetKw,
    IfKw,
    ElseKw,
    WhileKw,
    BreakKw,
    FunKw,
    TrueKw,
    FalseKw,
    Equals,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Number,
    OpenParen,
    CloseParen,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Root,
    FuncDef,
    ParamList,
    NumberLiteral,
    BoolLiteral,
    ArrayLiteral,
    BinaryExpr,
    PrefixExpr,
    ParenExpr,
    RefExpr,
    IfExpr,
    FnCallExpr,
    BlockExpr,
    Stmt,
    LetStmt,
    WhileStmt,
    BreakStmt,
    ExprStmt,
}

impl SyntaxKind {
    pub fn is_eof(self) -> bool {
        self == Self::Eof
    }

    pub fn is_error(self) -> bool {
        self == Self::Error
    }

    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::Whitespace)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}
