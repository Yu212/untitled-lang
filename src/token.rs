use rowan::TextRange;

use crate::syntax_kind::SyntaxKind;

pub struct Token<'a> {
    pub kind: SyntaxKind,
    pub text: &'a str,
    pub range: TextRange,
}
