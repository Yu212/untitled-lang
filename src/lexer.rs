use rowan::{TextRange, TextSize};
use unscanny::Scanner;

use crate::syntax_error::SyntaxError;
use crate::syntax_kind::SyntaxKind;
use crate::token::Token;

pub struct Lexer<'a> {
    s: Scanner<'a>,
    tokens: Vec<Token<'a>>,
    errors: Vec<SyntaxError>,
    error: Option<String>,
}

impl<'a> Lexer<'a> {
    pub fn new(text: &'a str) -> Self {
        Self {
            s: Scanner::new(text),
            tokens: Vec::new(),
            errors: Vec::new(),
            error: None,
        }
    }

    pub fn tokenize(mut self) -> (Vec<Token<'a>>, Vec<SyntaxError>) {
        loop {
            let start = self.s.cursor();
            let kind = self.identify_token();
            let end = self.s.cursor();
            let range = TextRange::new(TextSize::new(start as u32), TextSize::new(end as u32));
            let text = self.s.get(range.into());
            self.tokens.push(Token {
                kind,
                text,
                range,
            });
            if let Some(message) = self.error.take() {
                self.errors.push(SyntaxError {
                    message,
                    range,
                });
            }
            if kind.is_eof() {
                return (self.tokens, self.errors)
            }
        }
    }

    fn error(&mut self, message: impl Into<String>) -> SyntaxKind {
        self.error = Some(message.into());
        SyntaxKind::Error
    }

    fn identify_token(&mut self) -> SyntaxKind {
        match self.s.eat() {
            Some(c) if c.is_ascii_whitespace() => self.whitespace(),
            Some(c) if c.is_ascii_digit() => self.number(),
            Some('+') => SyntaxKind::Plus,
            Some('*') => SyntaxKind::Star,
            None => SyntaxKind::Eof,
            _ => self.error("Unexpected character"),
        }
    }

    fn whitespace(&mut self) -> SyntaxKind {
        self.s.eat_while(char::is_ascii_whitespace);
        SyntaxKind::Whitespace
    }

    fn number(&mut self) -> SyntaxKind {
        self.s.eat_while(char::is_ascii_digit);
        SyntaxKind::Number
    }
}
