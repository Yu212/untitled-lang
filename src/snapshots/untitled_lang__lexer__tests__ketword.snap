---
source: src/lexer.rs
expression: "tokenize(\"let\")"
---
(
    [
        Token {
            kind: LetKw,
            text: "let",
            range: 0..3,
        },
        Token {
            kind: Eof,
            text: "",
            range: 3..3,
        },
    ],
    [],
)
